metagraph/src/annotation/binary_matrix/row_diff/row_diff.cpp

```diff
@@ -1,118 +1,154 @@
+
+    for(uint64_t batch = 0; batch < batch_count; ++batch) {
+        std::vector<std::vector<uint64_t>> pred_buf(num_threads);
+        std::vector<std::vector<uint64_t>> succ_buf(num_threads);
+        std::vector<std::vector<bool>> pred_boundary_buf(num_threads);
+
+        #pragma omp parallel for num_threads(num_threads) schedule(static, 1)
+        for (uint32_t chunk = 0; chunk < std::max((uint32_t)1, num_threads); ++ chunk) {
+            uint64_t start = batch * batch_size + chunk * chunk_size + 1;
+            for (uint64_t i = start;
+                 i < std::min(graph.num_nodes() + 1, start + chunk_size); ++i) {
+                uint64_t boss_idx = graph.kmer_to_boss_index(i);
+                if (dummy[boss_idx] || terminal[boss_idx]) {
+                    succ_buf[chunk].push_back(0);
+                } else {
+                    const graph::boss::BOSS::TAlphabet w
+                            = boss.get_W(boss_idx) % boss.alph_size;
+                    uint64_t next = w ? graph.boss_to_kmer_index(boss.fwd(boss_idx, w)) : 0;
+                    succ_buf[chunk].push_back(next);
+                }
+                uint64_t back_idx = boss.bwd(boss_idx);
+                boss.call_incoming_to_target(
+                        back_idx, boss.get_W(back_idx), [&](BOSS::edge_index incoming_edge) {
+                          // terminal and dummy predecessors are ignored; also ignore
+                          // predecessors for which boss_idx is not the last outgoing
+                          // edge (bc. we only traverse the last outgoing at a bifurcation)
+                          if (terminal[incoming_edge] || dummy[incoming_edge]
+                                  || boss.fwd(incoming_edge,
+                                              boss.get_W(incoming_edge) % boss.alph_size)
+                                          != boss_idx)
+                              return;
+
+                          pred_buf[chunk].push_back(graph.boss_to_kmer_index(incoming_edge));
+                          pred_boundary_buf[chunk].push_back(0);
+                        });
+                pred_boundary_buf[chunk].push_back(1);
+            }
+        }
+        for (uint32_t i = 0; i < num_threads; ++i) {
+            for (uint32_t j = 0; j < succ_buf[i].size(); ++j) {
+                succ.push_back(succ_buf[i][j]);
+            }
+            for (uint32_t j = 0; j < pred_buf[i].size(); ++j) {
+                pred.push_back(pred_buf[i][j]);
+            }
+            for (uint32_t j = 0; j < pred_boundary_buf[i].size(); ++j) {
+                pred_boundary.push_back(pred_boundary_buf[i][j]);
+            }
+        }
+        ++progress_bar;
     }
+    succ.close();
+    pred.close();
+
+    std::ofstream fpred_boundary(outfbase + ".pred_boundary");
+    sdsl::rrr_vector rpred_boundary(to_sdsl(pred_boundary));
+    rpred_boundary.serialize(fpred_boundary);
+    fpred_boundary.close();
```

Comment:
This can be quite large, I think, why not use delimiters in `pred`? It will be about twice larger, but we read it with streaming which may be ok. What do you think?

Reply:
But I think at some point you convert it to rrr, which requires the full bit vector to be stored in RAM. This is 1 bit per k-mer -- unnecessary overhead.

Reply:
I had removed that conversion. Now it's just a plain old bit set.